server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkerPool.java

```diff
@@ -0,0 +1,137 @@
+    private final Executor executor;
+    private final Consumer<SnapshotShardContext> shardSnapshotter;
+    private final CheckedBiConsumer<SnapshotShardContext, FileInfo, IOException> fileSnapshotter;
+    private volatile int workerCount = 0;
+
+    public record SnapshotFileUpload(SnapshotShardContext context, FileInfo fileInfo, ActionListener<Void> listener) {}
+
+    public ShardSnapshotWorkerPool(
+        final int maxWorkers,
+        final Executor executor,
+        final Consumer<SnapshotShardContext> shardSnapshotter,
+        final CheckedBiConsumer<SnapshotShardContext, FileInfo, IOException> fileSnapshotter
+    ) {
+        assert maxWorkers > 0;
+
+        logger.info("starting shard snapshot worker pool of max size {}", maxWorkers);
+        this.maxWorkers = maxWorkers;
+        this.executor = executor;
+        this.shardSnapshotter = shardSnapshotter;
+        this.fileSnapshotter = fileSnapshotter;
+    }
+
+    public void enqueueShardSnapshot(final SnapshotShardContext context) {
+        logger.trace("enqueuing shard snapshot task [snapshotID={}, indexID={}]", context.snapshotId(), context.indexId());
+        synchronized (mutex) {
+            shardsToSnapshot.add(context);
+            ensureEnoughWorkers();
+        }
+    }
+
+    public void enqueueFileUpload(final SnapshotFileUpload snapshotFileUpload) {
+        logger.trace(
+            "enqueuing shard snapshot file upload task [snapshotID={}, indexID={}, file={}]",
+            snapshotFileUpload.context.snapshotId(),
+            snapshotFileUpload.context.indexId(),
+            snapshotFileUpload.fileInfo.name()
+        );
+        synchronized (mutex) {
+            filesToSnapshot.add(snapshotFileUpload);
+            ensureEnoughWorkers();
+        }
+    }
+
+    private void ensureEnoughWorkers() {
+        assert Thread.holdsLock(mutex);
+
+        int workersToCreate = Math.min(maxWorkers - workerCount, Math.max(filesToSnapshot.size(), shardsToSnapshot.size()));
+        if (workersToCreate > 0) {
+            logger.debug("starting {} shard snapshot workers", workersToCreate);
+        }
+        while (workersToCreate > 0) {
+            workerCount++;
+            startWorker(UUIDs.base64UUID());
+            workersToCreate--;
+        }
+        logger.debug("worker pool size is {}", workerCount);
+    }
+
+    private void workerDone() {
+        synchronized (mutex) {
+            assert workerCount > 0;
+
+            workerCount--;
+            if (workerCount < maxWorkers && (shardsToSnapshot.isEmpty() == false || filesToSnapshot.isEmpty() == false)) {
+                ensureEnoughWorkers();
+            }
+        }
+    }
+
+    public int size() {
+        return workerCount;
+    }
+
+    private void startWorker(final String workerId) {
+        logger.debug("starting snapshot pool worker {}", workerId);
+        executor.execute(() -> {
```

Comment:
Thank you both for the feedback.

> I think the answer here is "mostly no".

That's the part I am not entirely sure about when using two queues. I don't like the fact that it is not entirely predictable! It would be great to make sure there is a certain order regarding the sub-tasks, that will always be maintained when snapshotting. I am just not sure if it's an overkill, and don't want to add unnecessary code.

If it is not too much unnecessary code, I could use the `startTime` that is recorded in the `SnapshotsInProgress.Entry`. I think this is set once on master when the snapshot request gets processed, so it seems to stay unchanged afterwards. I could use a priority queue that sorts by snapshot `startTime` -> `snapshotId` -> task type(metadata work or file upload). I'm adding `startTime` to the mix, since ordering by `snapshotId` might still result in some weird corner case, where some snapshot keeps getting pushed back, due to ordering by  names/UUIDs. 

Would there be a downside to adding `startTime` to the ordering?

Reply:
> this only matters if snapshots are triggered at a rate that's higher than the rate at which they can be completed.

I don't think that's quite true - there can temporarily be a bit spike in snapshot work to do even if in the long run we will be able to get through it all. For instance if a hot node fails and a bunch of replicas get promoted then that's going to cause a one-off bigger snapshot job. By prioritising this work over the work for later-starting ones we would typically finish the big snapshot sooner. Snapshots that finish sooner are kind of more useful because of how that reduces the window of uncertainty about which data is/isn't included.

Reply:
> To me it seems preferable to just always do the meta work first, then the file uploads.

I think I agree here, but that means we should also check for new metadata work between file uploads right?